From a99173591a7d9f833ebd491c6a0e7d01f7ea5305 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Wed, 5 Jun 2024 17:21:13 +0200
Subject: [PATCH 627/841] fix(st-rif): use clr/set bit when configuring RIFSC

When configuring a peripheral with stm32_rifsc_ip_configure(), all the
other peripherals in the same RIFSC bank were changed their rights to
unsecure/unprivileged. Correct that with using mmio_setbits_32 or
mmio_clrbits_32 instead of mmio_write_32, depending on secure and
privileged rights wanted.

Use % instead of / to determine the bit corresponding to the peripheral
in the RIFSC registers.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 356069ca07b1e4fa78c6c81c06b186294271e40e fix(st-rif): use clr/set bit when configuring RIFSC]
Change-Id: Ifd952fdfec9db73ade56b108b70278eed1497d50
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/385557
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/st/rif/stm32_rifsc.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/st/rif/stm32_rifsc.c b/drivers/st/rif/stm32_rifsc.c
index 8f0463b7f..6783e5843 100644
--- a/drivers/st/rif/stm32_rifsc.c
+++ b/drivers/st/rif/stm32_rifsc.c
@@ -83,11 +83,11 @@ void stm32_rifsc_ip_configure(int rimu_id, int rifsc_id, uint32_t param)
 	assert(rifsc_id < STM32MP25_RIFSC_MAX_ID);
 #endif /* STM32MP25 */
 
-	bit = BIT(rifsc_id / U(32));
+	bit = BIT(rifsc_id % U(32));
 
 	/* Set peripheral accesses to Secure/Privilege only */
-	mmio_write_32(RIFSC_BASE + _RIFSC_RISC_SECCFGR(rifsc_id), bit);
-	mmio_write_32(RIFSC_BASE + _RIFSC_RISC_PRIVCFGR(rifsc_id), bit);
+	mmio_setbits_32(RIFSC_BASE + _RIFSC_RISC_SECCFGR(rifsc_id), bit);
+	mmio_setbits_32(RIFSC_BASE + _RIFSC_RISC_PRIVCFGR(rifsc_id), bit);
 
 	/* Apply specific configuration to RIF master */
 	mmio_write_32(RIFSC_BASE + _RIFSC_RIMC_ATTR(rimu_id), param);
-- 
2.47.2

